Unit tests for #2560

RiJo opened the pull request:
Test cases to reproduce #2560.

### Did you remember to?

- [X] Add test case(s)
- [ ] Update `CHANGES.txt`

We encourage pull requests that:

* Add new features to TestNG (or)
* Fix bugs in TestNG

If your pull request involves fixing SonarQube issues then we would suggest that you please discuss this with the 
[TestNG-dev](https://groups.google.com/forum/#!forum/testng-dev) before you spend time working on it.

RiJo committed:
Unit tests for #2560

juherr commented on core/src/test/java/test/factory/github2560/ConstructorTest.java:
```diff
@@ -0,0 +1,43 @@
+
+import org.testng.annotations.*;
+
+public class ConstructorTest {
```
By convention, suffix it by `Sample`: `ConstructorTestSample`

juherr commented on core/src/test/java/test/factory/github2560/FactoryTest.java:
```diff
@@ -0,0 +1,15 @@
+
+import org.testng.annotations.Factory;
+
+public class FactoryTest {
```
By convention, suffix it by `Sample`: `FactoryTestSample`

juherr commented on core/src/test/java/test/factory/github2560/TestClass.java:
```diff
@@ -0,0 +1,37 @@
+
+import org.testng.annotations.*;
+
+public class TestClass {
```
`TestClassSample`

RiJo committed:
Add Sample suffix to classes

RiJo committed:
WIP: example of relevant code changes

 - This commit gives some examples on a potential approach
   to solve this issue. It is only intended as a demo on
   which part of the code must be updated.

   Potentially, a new interface should be introduced to
   expose the additional functionality to avoid downcasting.

juherr commented:
Closing this as superseded by #3089, which fixed the same per-instance configuration-method handling reported in #2560/#3082.

The current implementation keeps configuration methods associated with the correct `@Factory` instance and the regression coverage in `ListenersTest` verifies `@BeforeClass`, `@BeforeMethod`, the test method, `@AfterMethod`, and `@AfterClass` for each instance.

The implementation proposed here was also explicitly marked as a WIP/example and is no longer appropriate for the current internals.

Thanks for the original reproducer and regression tests — they helped document the issue clearly.
